### PR TITLE
RefTerms in the header should depend on the whole referred body (#337)

### LINKED ISSUE #337: The type of a function can't depend on functions after it
This code currently tycks:

(in this code, no definition depends on any definitions after it)

```aya
open import Paths

public open data Unit | unit

def coe {a b : Type} (eq : a ↑ = b) (x : a) : b
  => ↑ arcoe eq.at x right


public open struct EventT : Type 1
  | AgentT : Type 0 => Unit

def GetterT => EventT -> Type 0
def SetterT => Type 0 -> EventT -> EventT
def Invariant (GetT : GetterT) (SetT : SetterT)
  => Pi (T : EventT) (E : Type 0) -> E ↑ = GetT (SetT E T)

public open struct Event (T : EventT) : Type 0
  | agent : T.AgentT

def Getter (GetT : GetterT) => Pi {T : EventT} (Event T) -> GetT T


def Quantifier (E : Type 0) : Type 1
  => (E -> Type 0) -> Type 0


-- verbs are event quantifiers
def Verb (T : EventT) : Type 1
  => (Event T -> Type 0) -> Type 0

-- VPs (verb phrases) behave just like verbs
def VP (T : EventT) => Verb T


-- bare DPs (determiner phrases) are just entity quantifiers
def BDP (E : Type 0) => Quantifier E

def DP (SetT : SetterT) (E : Type 0)
  => Pi {T : EventT} -> VP (SetT E T) -> VP (SetT E T)

-- turn a VP into a BDP
-- e.g. complementize GetAgentT getAgent =>
--   "Bob *killed Alice*." -> "*Who killed Alice* was Bob."
--def relativize (GetT : GetterT) (get : Getter GetT)
--  {T : EventT} (vp : VP T) :
def complementize (GetT : GetterT) (get : Getter GetT)
  {T : EventT} (vp : VP T) : BDP (GetT T)
  => \(f : GetT T -> Type 0) => vp (\(e : Event T) => f (get e))


-- thetas transform a BDP into a DP with that thematic role
def Theta (SetT : SetterT) : Type 1
  => Pi {E : Type 0} -> BDP E -> DP SetT E

def mkTheta (GetT : GetterT)
  (get : Getter GetT)
  (SetT : SetterT)
  (inv : Invariant GetT SetT)
  : Theta SetT
  => \{E : Type 0} (q : BDP E) -- from Theta
  => \{T : EventT} (p : VP (SetT E T)) -- from DP
  => \(f : Event (SetT E T) -> Type 0) -- from VP
  => q (\(x : E) => p (\(e : Event (SetT E T))
    => Sig (get e = coe (inv T E) x) ** (f e)))
```

After moving `complementize` before `BDP`, new errors arise in `complementize` **and `mkTheta`**(???).

I haven't found a minimal test yet.

### PR DESCRIPTION
**Pull request by imkiva**
This PR fixes #337

### CODE REVIEW

**Comment by imkiva** [comment]
bors merge

**Comment by imkiva** [comment]
bors cancel

**Comment by imkiva** [comment]
bors merge

File: base/src/main/java/org/aya/resolve/visitor/ExprResolver.java [hunk 1]
```diff
@@ -60,9 +60,12 @@ public void enterBody() {
   private void addReference(@NotNull TyckUnit unit) {
     if (parentAdd != null) parentAdd.accept(unit);
     if (where.isEmpty()) throw new IllegalStateException("where am I?");
-    var order = where.peek() == Where.Head
-      ? new TyckOrder.Head(unit) : new TyckOrder.Body(unit);
-    reference.append(order);
+    if (where.peek() == Where.Head) {
+      reference.append(new TyckOrder.Head(unit));
+      reference.append(new TyckOrder.Body(unit));
```
Comment (ice1000) [inline base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:65]: Ohhh!

Comment (imkiva) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:65]: I was stupid 😭

**Pull request merged**